### PR TITLE
add method to User for "Match Me".  Resolves #95

### LINKED ISSUE #95: New "match me" code
We're changing the way "Match me" works to direct a user to others who have answered at a non-"I want to learn" level on questions the user has answered "I want to learn" for.  See:

https://trello.com/c/6RqIpZxy/16-match-me-your-connectors

That's not the way "Match me" works right now.  We need a new methods for the user, each one returning a list of users, ideally within which there is a list of questions that they have answered at a specific level, specified to the method.  These questions themselves need to be grouped by topic.  Signature is:

```
user.match(level)
```

Where `level` is `LEVEL_I_CAN_EXPLAIN`, `LEVEL_I_CAN_DO_IT`, etc. etc.

### PR DESCRIPTION
**Pull request by talos**

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -218,6 +218,45 @@ def set_fully_registered(self):
             return
         db.session.add(UserJoinedEvent.from_user(self))
 
+    def match(self, level, limit=10):
+        '''
+        Return a list of users that can answer questions this user wants to
+        learn at the specified level.
```
Comment (toolness) [inline app/models.py]: Hmm, this says it's returning a list of users, but it seems like it's actually yielding tuples of users and ... um ... a list containing information about the categories/questions the user can answer, sorted by the number of questions in each category?

It kinda makes me wonder if there's an intermediate `MatchedUserResultSet` class or something we can factor out to make the return value a bit easier understand, but at the very least, documenting the return value a bit better would be really helpful!

**Comment by toolness** [comment]
Aside from the [complexity of the return value](https://github.com/GovLab/noi2/pull/97/files#r41327038) of the `match` method, I think this looks good!

On a side note, I'm a bit scared to tinker with that default fixture anymore--doing so would mean changing so many tests at this point!

File: app/models.py (hunk 1 above)
Comment (toolness) [inline app/models.py]: Alternatively, maybe the function could be refactored into smaller functions that are composed with one another? Might make it a bit easier to understand... or maybe I'm just dense, hehe.

**Comment by talos** [comment]
Yeah, the bundling of fixtures and tests at this point is problematic.  We need complicated skill values for users in order to be able to test certain aspects of the functionality, though.

What I'm leaning towards now would be test-specific fixtures, maybe even defined at the class-level.  What do you think?

File: app/models.py (hunk 1 above)
Comment (talos) [inline app/models.py]: Good points all.  I definitely should improve the documentation here.

The reason for the complexity of the result value is to make it easily consumable in a view.  Wrapping the complex structure of the response in some OO boilerplate ain't a bad idea.

File: app/utils.py [hunk 2]
```diff
@@ -15,3 +16,41 @@ def csv_reader(path_to_file):
...
+
+class UserSkillMatch(object):
+    """
+    An object linking a User with matched skills.
+
+    This object exposes `user` and `questionnaires` properties.  The
+    `questionnaires` property is a list of two-tuples, in descending order of
+    number of matched questions, where the first element is the name of the
+    questionnaire and the second element is a list (in no particular order) of
+    the the matched quesiton IDs.
```
Comment (toolness) [inline app/utils.py]: minor typo, `quesiton` -> `question`

Comment (toolness) [inline app/utils.py]: This is really great! The only thing that would make it even more rad would be really simple doctests to give the reader some usage examples, but that's really just extra icing on the cake.

**Comment by toolness** [comment]
> What I'm leaning towards now would be test-specific fixtures, maybe even defined at the class-level. What do you think?

Yeah, I definitely like that idea. There's also a funky thing called [factory_boy](https://factoryboy.readthedocs.org/en/latest/) that's intended as a replacement for fixtures... A Cucumber book I was reading was very enthusiastic about the ruby library it's based on, factory girl, so we could also give that a try if we want.

**Comment by talos** [comment]
Ah yes, I've heard good things about that library too!

**Pull request merged**